immuneML/ml_methods/PWM.py

```diff
@@ -0,0 +1,174 @@
+import csv
+import yaml
+import datetime
+
+import numpy as np
+import pandas as pd
+
+from pathlib import Path
+from immuneML.ml_methods.GenerativeModel import GenerativeModel
+from scripts.specification_util import update_docs_per_mapping
+from immuneML.util.PathBuilder import PathBuilder
+
+class PWM(GenerativeModel):
+
+    def get_classes(self) -> list:
+        pass
+
+    def __init__(self, parameter_grid: dict = None, parameters: dict = None):
+        parameters = parameters if parameters is not None else {}
+        parameter_grid = parameter_grid if parameter_grid is not None else {}
+        self._alphabet = ""
+        self.generated_sequences = []
+        super(PWM, self).__init__(parameter_grid=parameter_grid, parameters=parameters)
+
+    def _get_ml_model(self, cores_for_training: int = 2, dataset=None):
+
+
+        sequences = dataset.split(" ")
+        self._length_of_sequence = len(sequences[0])
+        sequences_without_spaces = "".join(sequences)
+        dataset = np.array(list(sequences_without_spaces))
+        dataset = np.reshape(dataset, (len(sequences), self._length_of_sequence))
+        self._alphabet = sorted(set("".join(sequences)))
+        matrix = np.zeros(shape=(self._length_of_sequence, len(self._alphabet)))
+
+        instances = dataset.T
+        for x, pos in enumerate(instances):
+            for element in pos:
+                for y, char in enumerate(self._alphabet):
+                    if element == char:
+                        matrix[x][y] += 1
+                        break
+
+        for ind, row in enumerate(matrix):
+            matrix[ind] = matrix[ind] / sum(matrix[ind])
+
+        return matrix
+
+    def _fit(self, X, cores_for_training: int = 1, result_path: Path = None):
+        self.model = self._get_ml_model(cores_for_training, X)
+
+        return self.model
+
+    def generate(self, length_of_sequences: int = None, amount=10, path_to_model: Path = None):
+        #
+        # if self.model is None:
+        #     model_as_array = []
+        #     print(f'{datetime.datetime.now()}: Fetching model...')
+        #     with open(path_to_model, 'r') as file:
+        #
+        #         reader = csv.reader(file)
+        #         self._alphabet = "".join(next(reader))
+        #         for row in reader:
+        #             model_as_array.append(row)
+        #     self.model = np.array(model_as_array)
+
+        # length_of_sequences = length_of_sequences if length_of_sequences is not None else self.model.shape[0]
+        length_of_sequences = self.model.shape[0]
+        generated_sequences = []
+        for _ in range(amount):
+            sequence = []
+            for i in range(length_of_sequences):
+                sequence.append(np.random.choice(self._alphabet, 1, p=self.model[i])[0])
+            generated_sequences.append(sequence)
+
+        instances = np.array(generated_sequences)
+
+        matrix = np.zeros(shape=(instances.shape[1], len(self._alphabet)))
```

Comment:
what is the aim of the matrix here? why not just generate sequences as strings (line 72) and export a list of those?